### PR TITLE
some smaller modernizations to the shutdown loop

### PR DESCRIPTION
**Pull request by poettering**
Some minor bugfixes and modernizations

### CODE REVIEW

**Comment by xnox** [comment]
This is nice. I like it.

But I think we also want to work against kernel repeatedly succeeding at clr_fd ioctl, without actually clearing fd and simply setting autoclear flag on, and not cleaning up backing_file at all.

See observations at https://github.com/systemd/systemd/issues/14410

**Comment by xnox** [comment]
Hm. I do wonder if we actually should call clr_fd ioctl on unmountable things at least once, as well. Such that kernel itself sets the autoclear flag on them.

**Comment by poettering** [comment]
> Hm. I do wonder if we actually should call clr_fd ioctl on unmountable things at least once, as well. Such that kernel itself sets the autoclear flag on them.

hmm, you have a point, we should set autoclear on devices that are still busy. will prep a fix.

**Comment by poettering** [comment]
I added another commit: when we get EBUSY from LOOP_CLR_FD we'll set LO_FLAGS_AUTOCLEAR, in the hope this cleans itself up eventually. ptal.

File: src/shutdown/umount.c [hunk 1]
```diff
@@ -325,23 +328,62 @@ static int dm_list_get(MountPoint **head) {
...
                 return 1;
+        }
 
-        /* ENXIO: not bound, so no error */
-        if (errno == ENXIO)
-                return 0;
+        /* Linux makes LOOP_CLR_FD succeed whenever LO_FLAGS_AUTOCLEAR is set without actually doing
+         * anything. Very confusing. Let's hence not claim we did anything in this case. */
+        if (FLAGS_SET(info.lo_flags, LO_FLAGS_AUTOCLEAR))
+                log_debug("Successfully called LOOP_CLR_FD on a loopback device %s with autoclear set, which is a NOP.", device);
```
Comment (xnox) [inline src/shutdown/umount.c:382]: This is slightly misleading.

ioctl can fail, set autoclear, or detach a device. Here we do a check of AUTOCLEAR flag, after calling the ioctl, so we don't know if autoclear was already set on it or if calling the ioctl did it.

I'd rather check flags before calling the clr_fd ioctl. If autoclear is on, and device is there, clearly it's busy, bail. Or call clr_fd, check flags again and then print "didn't do anything" or "set autoclear on".

File: src/shutdown/umount.c [hunk 2]
```diff
@@ -325,23 +328,62 @@ static int dm_list_get(MountPoint **head) {
...
+        else
+                log_debug("Weird, LOOP_CLR_FD succeeded but the device is still attached on %s.", device);
```
Comment (xnox) [inline src/shutdown/umount.c:384]: No it was simply no-autoclear set & busy, then clr_fd simply set the autoclear flag.

**Comment by xnox** [comment]
Actually, to make the shutdown quicker it would make sense to change our main shutdown loop slightly

1) set loopback autoclr flag on everything
2) then do unmount_all loop
3) then try to do clr_fd()
4) then do other things
5) loop

File: src/shutdown/umount.c [hunk 3]
```diff
@@ -325,23 +328,62 @@ static int dm_list_get(MountPoint **head) {
...
         if (fd < 0)
                 return errno == ENOENT ? 0 : -errno;
 
-        r = ioctl(fd, LOOP_CLR_FD, 0);
-        if (r >= 0)
+        if (ioctl(fd, LOOP_CLR_FD, 0) < 0) {
+                if (errno == ENXIO) /* Nothing bound, didn't do anything */
+                        return 0;
+
+                if (errno != EBUSY)
```
Comment (xnox) [inline src/shutdown/umount.c:343]: Since v3.7 kernel does not return EBUSY.

And we target v3.13+

**Comment by xnox** [comment]
> Actually, to make the shutdown quicker it would make sense to change our main shutdown loop slightly
> 
> 1. set loopback autoclr flag on everything
> 2. then do unmount_all loop
> 3. then try to do clr_fd()
> 4. then do other things
> 5. loop

this will not quite work. Setting autoclear doesn't cause device to attempt clr_fd, as that is only done when a device is released.

**Pull request merged**